sources/csv/data.go

```diff
@@ -0,0 +1,284 @@
+)
+
+type Column struct {
+	Column_name string `json:"column_name"`
+	Type_name   string `json:"type_name"`
+}
+
+// Harbourbridge accepts a manifest file in the form of a json which unmarshalls into the Table struct.
+type Table struct {
+	Table_name    string   `json:"table_name"`
+	File_patterns []string `json:"file_patterns"`
+	Columns       []Column `json:"columns"`
+}
+
+// LoadManifest reads the manifest file and unmarshalls it into a list of Table struct.
+// It also performs certain checks on the manifest.
+func LoadManifest(manifestFile string) ([]Table, error) {
+	manifest, err := ioutil.ReadFile(manifestFile)
+	if err != nil {
+		return nil, fmt.Errorf("can't read manifest file due to: %v", err)
+	}
+	tables := []Table{}
+	err = json.Unmarshal(manifest, &tables)
+	if err != nil {
+		return nil, fmt.Errorf("unable to unmarshall json due to: %v", err)
+	}
+	err = verifyManifest(tables)
+	if err != nil {
+		return nil, fmt.Errorf("manifest is incomplete: %v", err)
+	}
+	return tables, nil
+}
+
+// verifyManifest performs certain prechecks on the structure of the manifest.
+// Checks on valid file paths and empty CSVs are handled as conv.Unexpected errors later during processing.
+func verifyManifest(tables []Table) error {
+	if len(tables) == 0 {
+		return fmt.Errorf("no tables found")
+	}
+	for i, table := range tables {
+		name := table.Table_name
+		if name == "" {
+			return fmt.Errorf("table number %d (0-indexed) does not have a name", i)
+		}
+		if len(table.File_patterns) == 0 {
+			return fmt.Errorf("no file path provided for table %s", name)
+		}
+		cols := table.Columns
+		if len(cols) == 0 {
+			return fmt.Errorf("`columns` field for table %s is empty", name)
+		}
+		for j, col := range cols {
+			if col.Column_name == "" || col.Type_name == "" {
+				return fmt.Errorf("please provide column_name and type_name in `columns` field at position %d (0-indexed)", j)
+			}
+		}
+	}
+	return nil
+}
+
+// SetRowStats calculates the number of rows per table.
+func SetRowStats(conv *internal.Conv, tables []Table) {
+	for _, table := range tables {
+		for _, filePath := range table.File_patterns {
+			count, err := getCSVRowCount(filePath)
+			if err != nil {
+				conv.Unexpected(fmt.Sprintf("Couldn't get number of rows for table %s", table.Table_name))
+				continue
+			}
+			conv.Stats.Rows[table.Table_name] += count
+		}
+	}
+}
+
+// getCSVRowCount returns the number of data rows in the CSV file.
+func getCSVRowCount(filePath string) (int64, error) {
+	count := int64(0)
+	csvFile, err := os.Open(filePath)
+	if err != nil {
+		fmt.Printf("can't read csv file: %s due to: %v\n", filePath, err)
+	}
+	r := csvReader.NewReader(csvFile)
+	for {
+		_, err := r.Read()
+		if err == io.EOF {
+			break
+		}
+		if err != nil {
+			return 0, fmt.Errorf("can't read row from file: %s", filePath)
+		}
+		count++
+	}
+	// Exclude the first row, that is, column headers.
+	return count - 1, nil
```

Comment:
We should check for empty files here, else we might run into some edge cases

Reply:
Added checks, but for empty fiels, just reporting it in the report, not erroring out because thats what we do in other places currently.